File: app/schemas/rag.py
"""RAG 请求/响应 Schema"""

from datetime import datetime

from pydantic import BaseModel, Field


class SearchRequest(BaseModel):
    """知识库检索请求"""

    query: str = Field(description="检索查询文本")
    top_k: int = Field(default=5, ge=1, le=20, description="返回结果数量")
    mode: str = Field(default="hybrid", description="检索模式: bm25 | vector | hybrid")


class SearchResult(BaseModel):
    """单条检索结果"""

    chunk_id: int = Field(description="切片 ID")
    document_id: int = Field(description="所属文档 ID")
    file_name: str = Field(description="来源文件名")
    page_numbers: list[int] = Field(default=[], description="页码")
    heading_context: str = Field(default="", description="章节上下文")
    content: str = Field(description="原始内容（Markdown 表格 / LaTeX 公式 / 图片描述）")
    score: float = Field(description="Reranker 相关性分数 ∈ [0,1]，1.0 表示完全相关")


class SearchResponse(BaseModel):
    """检索结果列表"""

    results: list[SearchResult] = Field(description="检索结果")
    total: int = Field(description="结果数量")


class UploadResponse(BaseModel):
    """文档上传响应"""

    document_id: int = Field(description="文档 ID")
    file_name: str = Field(description="文件名")
    status: str = Field(default="processing", description="处理状态")


class DocumentListItem(BaseModel):
    """文档列表项"""

    id: int = Field(description="文档 ID")
    file_name: str = Field(description="文件名")
    file_path: str | None = Field(default=None, description="文件存储路径")
    file_size: int | None = Field(default=None, description="文件大小（字节）")
    file_ext: str | None = Field(default=None, description="文件后缀")
    status: str = Field(description="处理状态")
    chunk_count: int = Field(default=0, description="切片数量")
    retry_count: int = Field(default=0, description="重试次数")
    error_message: str | None = Field(default=None, description="错误信息")
    created_at: datetime | None = Field(default=None, description="上传时间")
    parse_started_at: datetime | None = Field(default=None, description="开始解析时间")
    parse_completed_at: datetime | None = Field(default=None, description="解析完成时间")
    updated_at: datetime | None = Field(default=None, description="更新时间")


class DocumentListResponse(BaseModel):
    """文档列表响应"""

    items: list[DocumentListItem] = Field(description="文档列表")
    total: int = Field(description="文档总数")


class DocumentDetail(BaseModel):
    """文档详情"""

    id: int = Field(description="文档 ID")
    file_name: str = Field(description="文件名")
    file_path: str | None = Field(default=None, description="文件存储路径")
    file_size: int | None = Field(default=None, description="文件大小（字节）")
    file_ext: str | None = Field(default=None, description="文件后缀")
    status: str = Field(description="处理状态")
    chunk_count: int = Field(default=0, description="切片数量")
    error_message: str | None = Field(default=None, description="错误信息")
    retry_count: int = Field(default=0, description="重试次数")
    created_at: datetime | None = Field(default=None, description="上传时间")
    parse_started_at: datetime | None = Field(default=None, description="开始解析时间")
    parse_completed_at: datetime | None = Field(default=None, description="解析完成时间")
    updated_at: datetime | None = Field(default=None, description="更新时间")


class ChunkItem(BaseModel):
    """单条切片信息"""

    id: int = Field(description="切片 ID")
    document_id: int = Field(description="所属文档 ID")
    file_name: str = Field(description="文件名称")
    page_numbers: list[int] = Field(default=[], description="页码")
    heading_context: str = Field(default="", description="章节上下文")
    raw_content: str = Field(default="", description="原始内容（Markdown / LaTeX / 图片描述）")


class ChunkListResponse(BaseModel):
    """切片列表响应"""

    items: list[ChunkItem] = Field(description="切片列表")
    total: int = Field(description="切片总数")


class DeleteResponse(BaseModel):
    """删除响应"""

    deleted_chunks: int = Field(description="删除的切片数量")
    file_name: str = Field(description="文件名")

File: web/src/api/rag.ts
import request from './request'

export interface DocumentItem {
  id: number
  file_name: string
  file_size: number | null
  status: string
  error_message: string | null
  chunk_count: number
  page_numbers: number[]
  created_at: string | null
  parse_started_at: string | null
  parse_completed_at: string | null
}

export interface SearchResult {
  chunk_id: number
  file_name: string
  page_numbers: number[]
  heading_context: string
  content: string
  score: number
}

// 上传文档
export function uploadDocument(file: File) {
  const form = new FormData()
  form.append('file', file)
  return request.post<{ code: number; msg: string; data: { file_name: string; status: string } }>(
    '/document/upload',
    form,
  )
}

// 获取文档列表
export function getDocuments() {
  return request.get<{ code: number; msg: string; data: { items: DocumentItem[]; total: number } }>(
    '/document/',
  )
}

// 删除文档
export function deleteDocument(documentId: number) {
  return request.delete<{ code: number; msg: string; data: { deleted_chunks: number; file_name: string } }>(
    `/document/${documentId}`,
  )
}

// 取消解析
export function cancelDocument(documentId: number) {
  return request.post(`/document/${documentId}/cancel`)
}

// 重试解析
export function retryDocument(documentId: number) {
  return request.post(`/document/${documentId}/retry`)
}

// 下载原始文件
export function getDownloadUrl(documentId: number): string {
  return `/api/v1/document/${documentId}/download`
}

// 文档切片
export interface ChunkItem {
  id: number
  document_id: number
  page_numbers: number[]
  heading_context: string
  raw_content: string
}

export function getDocumentChunks(documentId: number) {
  return request.get<{ code: number; msg: string; data: { items: ChunkItem[]; total: number } }>(
    `/document/${documentId}/chunks`,
  )
}

// 文档每页解析结果
export interface PageData {
  page_no: number
  width: number
  height: number
  markdown: string
  table_count: number
  picture_count: number
}

export function getDocumentPages(documentId: number) {
  return request.get<{ code: number; msg: string; data: { pages: PageData[]; total: number } }>(
    `/document/${documentId}/pages`,
  )
}

// 知识库检索
export function searchKnowledge(query: string, topK = 5, mode = 'hybrid') {
  return request.post<{ code: number; msg: string; data: { results: SearchResult[]; total: number } }>(
    '/rag/search',
    { query, top_k: topK, mode },
  )
}
